File: ts-writer-workspace/iteration-1/eval-1-follow-patterns/without_skill/outputs/src/services/projectService.ts
import { db } from '../db';
import { NotFoundError, ValidationError } from '../errors';

type ProjectStatus = 'active' | 'archived';

interface Project {
  id: string;
  name: string;
  description: string;
  ownerId: string;
  status: ProjectStatus;
  createdAt: Date;
  updatedAt: Date;
}

interface CreateProjectInput {
  name: string;
  description: string;
  ownerId: string;
}

interface UpdateProjectInput {
  name?: string;
  description?: string;
}

export class ProjectService {
  async create(input: CreateProjectInput): Promise<Project> {
    if (!input.name.trim()) throw new ValidationError('Name cannot be empty');
    if (!input.description.trim()) throw new ValidationError('Description cannot be empty');

    const owner = await db.users.findById(input.ownerId);
    if (!owner) throw new NotFoundError('User', input.ownerId);

    const now = new Date();
    return db.projects.create({
      name: input.name.trim(),
      description: input.description.trim(),
      ownerId: input.ownerId,
      status: 'active' as ProjectStatus,
      createdAt: now,
      updatedAt: now,
    });
  }

  async getById(id: string): Promise<Project> {
    const project = await db.projects.findById(id);
    if (!project) throw new NotFoundError('Project', id);
    return project;
  }

  async listByOwner(ownerId: string): Promise<Project[]> {
    return db.projects.findByUserId(ownerId);
  }

  async update(id: string, input: UpdateProjectInput): Promise<Project> {
    const project = await db.projects.findById(id);
    if (!project) throw new NotFoundError('Project', id);
    if (project.status === 'archived') throw new ValidationError('Cannot update an archived project');

    if (input.name !== undefined && !input.name.trim()) throw new ValidationError('Name cannot be empty');
    if (input.description !== undefined && !input.description.trim()) throw new ValidationError('Description cannot be empty');

    return db.projects.update(id, {
      ...(input.name !== undefined && { name: input.name.trim() }),
      ...(input.description !== undefined && { description: input.description.trim() }),
      updatedAt: new Date(),
    });
  }

  async archive(id: string): Promise<Project> {
    const project = await db.projects.findById(id);
    if (!project) throw new NotFoundError('Project', id);
    if (project.status === 'archived') throw new ValidationError('Project is already archived');

    return db.projects.update(id, {
      status: 'archived' as ProjectStatus,
      updatedAt: new Date(),
    });
  }
}

File: py-writer-workspace/iteration-1/eval-1-follow-patterns/with_skill/outputs/app/services/project_service.py
from app.models.project import Project, ProjectStatus
from app.repositories.base import ProjectRepository, UserRepository
from app.exceptions import NotFoundError, ValidationError


class ProjectService:
    def __init__(self, repo: ProjectRepository, user_repo: UserRepository) -> None:
        self._repo = repo
        self._user_repo = user_repo

    async def get_by_id(self, project_id: str) -> Project:
        project = await self._repo.find_by_id(project_id)
        if project is None:
            raise NotFoundError("Project", project_id)
        return project

    async def list_by_owner(self, owner_id: str) -> list[Project]:
        return await self._repo.find_by_owner(owner_id)

    async def create(self, name: str, description: str, owner_id: str) -> Project:
        owner = await self._user_repo.find_by_id(owner_id)
        if owner is None:
            raise NotFoundError("User", owner_id)
        return await self._repo.create(
            name=name,
            description=description,
            owner_id=owner_id,
            status=ProjectStatus.ACTIVE,
        )

    async def update(self, project_id: str, name: str | None = None, description: str | None = None) -> Project:
        project = await self._repo.find_by_id(project_id)
        if project is None:
            raise NotFoundError("Project", project_id)
        if name is not None:
            project.name = name
        if description is not None:
            project.description = description
        return await self._repo.update(project)

    async def archive(self, project_id: str) -> Project:
        project = await self._repo.find_by_id(project_id)
        if project is None:
            raise NotFoundError("Project", project_id)
        if project.status == ProjectStatus.ARCHIVED:
            raise ValidationError(f"Project already archived: {project_id}")
        project.status = ProjectStatus.ARCHIVED
        return await self._repo.update(project)
